### PR TITLE
Fixes #10521 - reporting templates

### PR DESCRIPTION
**Pull request by ares**
reporting

This is an initial version that does the rendering in web process. It comes with only one built-in report. It also provides API, hammer support will be trivial.

In future but not in this PR, I'd like to add 
1) template inputs similar to how it works with JobTemplates, that's currently blocked by https://github.com/theforeman/foreman/pull/5545
2) scheduling generating to future
3) offloading to dynflow executor
4) enrich the template DSL with more helpers for commonly asked reports, implement more reports
5) upload report templates to community templates repo (blocked by merging this PR)
6) verify foreman_templates works, should work out of the box though
7) write a chapter in manual (blocked by merging this PR)
8) add support to hammer

For those who discussed the original design, I came to conclusion that streaming the rendered template is not possible if we want to leverage power of erb for rendering itself. Loading of objects should be handled by new macros (see `load_hosts`) that ensures we process objects in batches. Rendering will be shortly possible asynchronously by dynflow, we'll probably deliver the result via email in such case, so for big reports, web process rendering/preview will be only useful for reasonable sizes.

### CODE REVIEW

**Comment by ares** [comment]
cc @tstrachota, please take a look when you have a chance, I'd appreciate help with steps 1. and 2. where I'd like to implement the scheduling form using react

File: app/controllers/report_templates_controller.rb
Comment (timogoebel) [inline app/controllers/report_templates_controller.rb:10]: Nit: I'd use a before_action for this.

Comment (ares) [reply app/controllers/report_templates_controller.rb:10]: I wanted too, but it's inherited from templates_controller, if I specify it here, I'd need to copy all listed actions from templaters_controller, adding just generate action here overrides the whole list. I could extract the list of action to some constant/method but this felt slightly more straightforward.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/report_templates_controller.rb]: Why is this needed?

File: lib/foreman/renderer.rb
Comment (timogoebel) [inline lib/foreman/renderer.rb]: I believe we should not add this here. Does it make sense to have this method e.g. in a provisioning template? This method will be available in the host object as well. `Host.first.load_hosts` does not make a lot of sense. Luckily #5683 will provide an easy way to make this possible.

**Review by timogoebel — COMMENTED** [review]
Not a full review, just some comments I had while reading the code.

**Comment by timogoebel** [comment]
> offloading to dynflow executor

What would be gain by rendering the template in the dynflow executor?

File: app/controllers/api/v2/report_templates_controller.rb
Comment (tstrachota) [inline app/controllers/api/v2/report_templates_controller.rb]: Copied over by mistake?

File: app/controllers/api/v2/report_templates_controller.rb
Comment (tstrachota) [inline app/controllers/api/v2/report_templates_controller.rb]: `"Import a report template"`

File: app/controllers/api/v2/report_templates_controller.rb
Comment (tstrachota) [inline app/controllers/api/v2/report_templates_controller.rb]: `s/partition template/report template`

File: app/views/report_templates/_diff.html.erb
Comment (tstrachota) [inline app/views/report_templates/_diff.html.erb]: Weird spacing:
` "diffLines","" ,:size `
vs.
` "diffLines", "", :size `
and also add space before `%>` please

**Review by tstrachota — CHANGES_REQUESTED** [review]
I haven't done any deep testing yet. Sending just first batch of small issues I came across while reading the code.

File: lib/foreman/renderer.rb
Comment (ares) [reply lib/foreman/renderer.rb]: I agree, once the refactoring is done, I'd create a new scope for reporting templates (or other ways around, based on what gets in first)

**Comment by ares** [comment]
fixed/replied to all comments I received so far, one question raised by @timogoebel 

> What would be gain by rendering the template in the dynflow executor?

Originally I thought we'd use the same workflow for "render now" and "render later" and potentially the web process wouldn't be blocked in case the report is rendered for a while, dynflow executor can run not only as a separate process but also on a separate host. I realized this would be quite complicated with ActiveJob today, so offloading will be possible only with delayed rendering. Luckily we can use ActiveJob with dynflow backed for this now.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (ares) [reply app/controllers/api/v2/report_templates_controller.rb]: leftover, removed

File: app/controllers/api/v2/report_templates_controller.rb
Comment (tstrachota) [inline app/controllers/api/v2/report_templates_controller.rb]: I guess you should use only 'generate' here.

File: app/controllers/report_templates_controller.rb
Comment (tstrachota) [inline app/controllers/report_templates_controller.rb]: Same here, 'generate' is the right value.

**Review by tstrachota — CHANGES_REQUESTED** [review]
Thanks for your updates @ares. The code looks good to me.

I tested the functionality and I have troubles with rendering reports as a regular user (not an admin). I always get 404 `Resource report_template not found by id '110'` even though I can view and edit the template normally. I have the "generate" permission too. 

After some investigation I found out that the permission created in `scope_for` is `"generate_report_templates_report_templates"`. Both UI and API controllers are affected.

**Comment by tstrachota** [comment]
@ares I'm little concerned about the preview functionality for large reports. If I understand it correctly it triggers rendering on each 'Preview' click. Won't it be too much of unexpected burden?

**Comment by ares** [comment]
> @ares I'm little concerned about the preview functionality for large reports. If I understand it correctly it triggers rendering on each 'Preview' click. Won't it be too much of unexpected burden?

I agree with the concern, I'm waiting on templates refactoring to be merged to introduce also rendering mode (real/preview). In preview mode we could limit all resource loader with `.limit(10)` for example.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (ares) [reply app/controllers/api/v2/report_templates_controller.rb]: fixed

File: app/controllers/report_templates_controller.rb
Comment (ares) [reply app/controllers/report_templates_controller.rb]: fixed

**Comment by ares** [comment]
Thanks for thorough testing under non-admin user. I pushed the permission change into last commit, please take another look.

**Comment by ares** [comment]
rebased after medium rendering was merged

**Comment by tstrachota** [comment]
Thanks @ares! Some tests are failing. Apart from that it's good to go.

**Comment by ares** [comment]
2 failures should be fixed in last commit, 3rd is unrelated and this should fix it https://github.com/theforeman/foreman/pull/5882

**Comment by ares** [comment]
rebased to make Foreman tests green, squashed as asked

**Comment by tstrachota** [comment]
Thanks, waiting for results of the tests to merge.

**Comment by timogoebel** [comment]
> Thanks, waiting for results of the tests to merge.

@ares said he already has a branch that rebased on the template scope PR. Let's wait for the other PR to get in before we merge this. I can hit the button next week, no problem.

File: lib/foreman/renderer/scope/macros/base.rb
Comment (timogoebel) [inline lib/foreman/renderer/scope/macros/base.rb:81]: Do we want to have this in the base scope and available in all templates? I think it makes sense to move this to a dedicated scope for reports.

File: test/unit/foreman/renderer/renderers_shared_tests.rb
Comment (timogoebel) [inline test/unit/foreman/renderer/renderers_shared_tests.rb:141]: I'd move these tests to a dedicated scope class for report templates.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ares: Thanks, I have two more suggestions after the rebase.

File: lib/foreman/renderer/scope/macros/base.rb
Comment (ares) [reply lib/foreman/renderer/scope/macros/base.rb:81]: I thought this would be useful everwhere, e.g. in host's kickstart template, you may want to render a config file based on data of other hosts, e.g. setting up a gateway, /etc/hosts etc

File: test/unit/foreman/renderer/renderers_shared_tests.rb
Comment (ares) [reply test/unit/foreman/renderer/renderers_shared_tests.rb:141]: same as the previous one

File: app/controllers/templates_controller.rb
Comment (timogoebel) [inline app/controllers/templates_controller.rb:113]: 😢 This method had a short life :-(
We should ideally keep this as unattended controller still uses the method.

Comment (timogoebel) [reply app/controllers/templates_controller.rb:113]: In addition, the reporting template should probably use it's own scope.
Maybe we need to use `template.get_scope` or something...

**Review by timogoebel — CHANGES_REQUESTED** [review]

File: app/controllers/templates_controller.rb
Comment (ares) [reply app/controllers/templates_controller.rb:113]: my exact thoughts after our IRC discussion, see the latest commit, I couldn't resist :-) we can keep it for now or I can do the similar change there if you prefer

**Review by tstrachota — DISMISSED** [review]

**Comment by tstrachota** [comment]
It works good and the design seems fine to me either, but I'll leave the final word on @timogoebel since he has a lot more insights into the latest template refactoring.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/report_templates_controller.rb]: Luckily, we don't need this anymore.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (timogoebel) [inline app/controllers/api/v2/report_templates_controller.rb]: Why do we need this? `load_vars_from_template` is not used here anywhere.

File: app/controllers/templates_controller.rb
Comment (timogoebel) [reply app/controllers/templates_controller.rb:113]: I like the default class defined for every template type. 

How about we introduce a `template.render` method that looks similar to this:

```
def render(template:, mode: Foreman::Renderer::REAL_MODE, host: nil, params: {})
  source = Foreman::Renderer.get_source(template: self, host: host)
  scope = Foreman::Renderer.get_scope(host: host, params: params, mode: mode, template: self)
  Foreman::Renderer.render(source, scope)
end
```

That could replace host.render_template but we dry this up a lot more.

File: lib/foreman/renderer/base_renderer.rb
Comment (timogoebel) [inline lib/foreman/renderer/base_renderer.rb]: I might be a little overly strict here, but is there a way we can use inheritance for this?
Maybe `if !scope.previw && template.log_results?` ?

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ares: I went over the code one more time with more focus. Left some comments.
In general: Nice, we're close.

File: app/controllers/api/v2/report_templates_controller.rb
Comment (ares) [reply app/controllers/api/v2/report_templates_controller.rb]: no longer needed, I know that without this, renderer used to fail since it tried to call `load_vars_from_template`, which is why we also had this method empty here, it works fine after refactoring

**Comment by ares** [comment]
I hope I addressed all comments, I like `template.render` idea, it deduplicated the code a lot. Also we no longer need provisioning_template render_template method.

**Review by timogoebel — APPROVED** [review]
LGTM, APJ.

File: app/controllers/templates_controller.rb
Comment (timogoebel) [inline app/controllers/templates_controller.rb]: @ares: Can you please also change this in the unattended controller to be consistent? Thanks.

**Review by timogoebel — CHANGES_REQUESTED** [review]
I found one small nit, otherwise still 👍.

File: lib/foreman/renderer.rb
Comment (kamils-iRonin) [inline lib/foreman/renderer.rb:35]: Maybe we can pass `klass` here, instead of `template` instance?

Comment (ares) [reply lib/foreman/renderer.rb:35]: sorry, not sure I follow, if we pass klass as an argument, it wins `klass ||= `, then we default to template's default class and then we fallback to Foreman::Renderer::Scope::Provisioning

File: app/controllers/templates_controller.rb
Comment (ares) [reply app/controllers/templates_controller.rb]: pushed in separate commit, good point

**Comment by ares** [comment]
@timogoebel tests are green :-)

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
Thanks, @ares.

**Pull request merged**
